Handle error

sebing opened the pull request:
Fixing issue
https://github.com/yegor256/takes/issues/522

davvd commented:
@sebing Thanks, someone will review your pull request soon

davvd commented:
@pinaf it's yours,please review

pinaf commented on src/main/java/org/takes/facets/auth/social/PsGithub.java:
```diff
@@ -153,7 +153,7 @@ private String token(final String home, final String code)
             .path("login").path("oauth").path("access_token")
             .toString();
-        return new JdkRequest(uri)
+        final XmlResponse xmlres = new JdkRequest(uri)
```
@sebing let's name this variable `response`.

pinaf replied:
@sebing let's keep the return here and move the `try` to the line above.

pinaf commented on src/main/java/org/takes/facets/auth/social/PsGithub.java:
```diff
@@ -164,8 +164,16 @@ private String token(final String home, final String code)
+            .as(XmlResponse.class);
+        String tokenString = null;
+        try {
+            tokenString = xmlres.xml()
```
@sebing let's move the `return` statement here.

pinaf commented on src/main/java/org/takes/facets/auth/social/PsGithub.java:
```diff
@@ -164,8 +164,16 @@ private String token(final String home, final String code)
-            .as(XmlResponse.class)
-            .xml().xpath("/OAuth/access_token/text()").get(0);
+            .as(XmlResponse.class);
+        String tokenString = null;
```
@sebing and get rid of this variable.

pinaf commented on src/main/java/org/takes/facets/auth/social/PsGithub.java:
```diff
@@ -180,7 +188,7 @@ private static Identity parse(final JsonObject json) {
         props.put("login", json.getString("login", "unknown"));
         props.put("avatar", json.getString("avatar_url", "#"));
         return new Identity.Simple(
-            String.format("urn:github:%d", json.getInt("id")), props
```
@sebing please revert this change

pinaf commented:
@sebing we need a unit test for the new behavior.

pinaf commented on src/main/java/org/takes/facets/auth/social/PsGithub.java:
```diff
@@ -164,8 +164,16 @@ private String token(final String home, final String code)
+                .xpath("/OAuth/access_token/text()").get(0);
+        } catch (final IndexOutOfBoundsException exce) {
+            throw new HttpException(
+                HttpURLConnection.HTTP_BAD_REQUEST,"Invalid XML",exce);
```
@sebing please add whitespaces after each `,`

pinaf replied:
@sebing also, let's move the ending `);` to the next line, aligned with the `throw`.

sebing commented:
@pinaf thanks for the suggestions, I corrected them. 
I created the unit test case for the changes too. 

Please verify and let me know your views.

pinaf commented on src/test/java/org/takes/facets/auth/social/PsGithubTest.java:
```diff
@@ -148,4 +176,29 @@ private static void assertParam(final Request req,
+    /**
+     * An inner class for the Take implementation testing.
+     */
+    private static class PsGithubTestTake implements Take {
```
@sebing why create an inner class if it's only used in 1 place?

sebing replied:
Anonymous inner class max allowed is 20

pinaf replied:
@sebing would also be a good idea to rename the class to `TkFakeLogin` or something similar.

sebing replied:
@pinaf ( i will push the change with other changes)

pinaf replied:
@sebing also, please add `final` to this class.

sebing replied:
@pinaf done.

pinaf commented on src/test/java/org/takes/facets/auth/social/PsGithubTest.java:
```diff
@@ -135,6 +152,17 @@ public void exec(final URI home) throws IOException {
+     * without the access token.
+     * @return A basic directive.
+     */
+    private static Directives getDirectiveWithoutAccessToken() {
```
@sebing let's rename this to `basicDirectives`.

sebing replied:
@pinaf  I thought it was very specific directive without the access token, If we rename it to basicDirectives, tomorrow people may confuse?

pinaf replied:
@sebing I see your point. Let's remove the `get` though. `directivesWithoutAccessToken` would be nice.

pinaf commented on src/test/java/org/takes/facets/auth/social/PsGithubTest.java:
```diff
@@ -135,6 +152,17 @@ public void exec(final URI home) throws IOException {
     }
 
     /**
+     * A private method to create the basic directive.
```
@sebing basic directive`s, without access token.`

sebing replied:
Corrected the mistake.

pinaf commented on src/test/java/org/takes/facets/auth/social/PsGithubTest.java:
```diff
@@ -54,12 +57,51 @@
+    private static final String GIT_HUB_TOKEN = "GitHubToken";
+
+    /**
+     * A constant for access_token.
```
@sebing isn't this a header? If so, let's go with `Access token HTTP header.`

pinaf replied:
@sebing this one still needs change.

sebing replied:
@pinaf Added the comment with XPath.

pinaf commented on src/test/java/org/takes/facets/auth/social/PsGithubTest.java:
```diff
@@ -54,12 +57,51 @@
  */
 public final class PsGithubTest {
     /**
+     * A constant for GitHubToken.
```
@sebing `GitHubToken`

pinaf commented on src/test/java/org/takes/facets/auth/social/PsGithubTest.java:
```diff
@@ -54,12 +57,51 @@
+     * @throws IOException If some problem inside.
+     */
+    @Test
+    public void throwHttpException() throws IOException {
```
@sebing this needs a better name. perhaps `failsOnNoAccessToken`

sebing replied:
Corrected.

pinaf commented on src/test/java/org/takes/facets/auth/social/PsGithubTest.java:
```diff
@@ -54,12 +57,51 @@
+    public transient ExpectedException thrown = ExpectedException.none();
+
+    /**
+     * PsGithub can throw HttpExeption.
```
@sebing PsGithub can `fail when no access token is present`.

pinaf commented on src/main/java/org/takes/facets/auth/social/PsGithub.java:
```diff
@@ -153,19 +153,25 @@ private String token(final String home, final String code)
+                .xml().xpath("/OAuth/access_token/text()").get(0);
+        } catch (final IndexOutOfBoundsException exception) {
+            throw new HttpException(
+                HttpURLConnection.HTTP_BAD_REQUEST, "Invalid XML", exception
```
@sebing let's change the error message here to `No access token` instead of the generic `Invalid XML`.

pinaf commented:
@sebing a few more comments and we are good.

sebing commented:
@pinaf I corrected the suggestions. (except renaming the method to  basic Directive)

pinaf commented on src/test/java/org/takes/facets/auth/social/PsGithubTest.java:
```diff
@@ -148,4 +175,29 @@ private static void assertParam(final Request req,
     }
+
+    /**
+     * An inner class for the Take implementation testing.
```
@sebing in that case let's change this to something more detailed. How about
`A test Take that verifies if the request contains the access token and returns a JSON login response` ?

pinaf commented on src/test/java/org/takes/facets/auth/social/PsGithubTest.java:
```diff
@@ -135,6 +152,16 @@ public void exec(final URI home) throws IOException {
     }
 
     /**
+     * A private method to create the basic directives, without access token.
```
@sebing no need to mention `A private method to...`. `Creates basic directives without an access token.` is fine.

sebing replied:
@pinaf Corrected.

pinaf commented on src/test/java/org/takes/facets/auth/social/PsGithubTest.java:
```diff
@@ -54,12 +57,51 @@
+    public transient ExpectedException thrown = ExpectedException.none();
+
+    /**
+     * PsGithub can fails on no access token.
```
@sebing can fail, without the `s`.

sebing replied:
@pinaf Corrected.

pinaf commented on src/test/java/org/takes/facets/auth/social/PsGithubTest.java:
```diff
@@ -54,12 +57,51 @@
+    }
+
+    /**
+     * A private method to perform the basic login.
```
@sebing `Performs a basic login` is fine.

sebing replied:
@pinaf Corrected.

pinaf commented on src/test/java/org/takes/facets/auth/social/PsGithubTest.java:
```diff
@@ -54,12 +57,51 @@
+     * A private method to perform the basic login.
+     * @param directive The directive object.
+     * @throws IOException If some problem inside.
+     * @checkstyle MultipleStringLiteralsCheck (100 lines)s
```
@sebing do we really need this? also, there is an extra `s` at the end.

sebing replied:
@pinaf Removed the line. (this line was there in the code already).

pinaf commented on src/test/java/org/takes/facets/auth/social/PsGithubTest.java:
```diff
@@ -54,12 +57,50 @@
+     * XPath element access_token.
+     */
+    private static final String ACCESS_TOKEN = "access_token";
+    /**
```
@sebing let's add a blank line here for consistency.

pinaf commented on src/test/java/org/takes/facets/auth/social/PsGithubTest.java:
```diff
@@ -135,6 +151,16 @@ public void exec(final URI home) throws IOException {
+     * Creates the basic directives, without access token.
+     * @return A basic directive.
+     */
+    private static Directives getDirectiveWithoutAccessToken() {
```
@sebing Let's remove the get. `directivesWithoutAccessToken` would be nice.

sebing commented:
@pinaf 

> The javadoc is can fail but the method is fails.

I thought there should be consistancy.

pinaf commented:
@sebing there is - it's just english. the test "fails", but the test "can fail" :)

sebing commented:
@pinaf I corrected the suggestions .

pinaf commented:
@rultor merge

pinaf commented:
@sebing good job! thanks!

rultor commented:
> @rultor merge

@pinaf Thanks for your request. @yegor256 Please confirm this.

sebing commented:
@pinaf  Thanks again for the suggestions and help.

sebing commented:
@yegor256 @davvd Any update on the closure of this merge?

yegor256 commented on src/main/java/org/takes/facets/auth/social/PsGithub.java:
```diff
@@ -153,19 +153,25 @@ private String token(final String home, final String code)
+                .assertStatus(HttpURLConnection.HTTP_OK)
+                .as(XmlResponse.class)
+                .xml().xpath("/OAuth/access_token/text()").get(0);
+        } catch (final IndexOutOfBoundsException exception) {
```
@sebing this looks ugly. should be like this instead:

```
List<String> tokens = xpath();
if (tokens.isEmpty()) {
  throw...
}
return tokens.get(0);
```

sebing replied:
@yegor256 The problem is the call to xpath is throwing the exception, so it won't reach the tokens.isEmpty line.

- I would like to have the return statement in the last line of the method, but @pinaf suggested this is the better way

yegor256 replied:
@sebing what exception you're talking about?

sebing replied:
@yegor256 The actual problem, we are trying to fix.
```
com.jcabi.xml.ListWrapper$NodeNotFoundException: XPath '/OAuth/access_token/text()' not found in
```
Is thrown in the call the `XML.xpath()` method.

yegor256 replied:
@sebing it's thrown only if you call `get()`, but not `isEmpty()`

yegor256 commented:
@sebing see my comments above

sebing commented:
@yegor256 I refactored the code. Can you have a look at them, and accept them?

yegor256 commented:
@rultor try to merge

rultor commented:
> @rultor try to merge

@yegor256 OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/6683-179516521)

rultor commented:
> @rultor try to merge

@yegor256 Done! FYI, the full log is [here](http://www.rultor.com/t/6683-179516521) (took me 18min)

davvd commented:
@ypshenychka please, check this issue for QA compliance, as per [par.24](http://at.teamed.io/policy.html#24)

davvd commented:
@rultor please deploy

rultor commented:
> @rultor please deploy

@davvd OK, I'll try to deploy now. You can check the progress [here](http://www.rultor.com/t/6683-180389597)

ypshenychka commented:
@sebing 
According to our [QA Rules](http://at.teamed.io/qa.html):
>Pull request title explains the problem it is fixing.
Pull request description explains the solution proposed and contains a link to the original ticket it is related to.

Please correct title and description according to above rules.

ypshenychka commented:
@sebing 
According to our [QA Rules](http://at.teamed.io/qa.html):
>Messages in a ticket always start with a name of a user they are addressed to.

Please correct couple of your code review comments by indicating an addressee in the beginning.

rultor commented:
> @rultor please deploy

@davvd Done! FYI, the full log is [here](http://www.rultor.com/t/6683-180389597) (took me 13min)

sebing commented:
@ypshenychka 
- updated the PR title.
- updated the PR description.
- added missing addressee in the comments. 

Thanks for the suggestions. I will try to be more careful next time :)

ypshenychka commented:
@sebing One more left :)
>sebing added a note 5 days ago
Corrected with a new name.

sebing commented:
@ypshenychka sorry for missing out one . I corrected them.

ypshenychka commented:
@sebing Thank you.

ypshenychka commented:
@davvd Quality is acceptable.

davvd commented:
> @davvd Quality is acceptable.

@ypshenychka thank you, let's try to make it "good" next time

davvd commented:
@pinaf 10 mins was added to the account of @ypshenychka (for QA review), in transaction 76577111; I added **44 mins** to your account, many thanks for working with the project (AP-45Y90720W83226110)! The completion time here was 142 hours and 50 mins.; extra minutes for review comments (c=29); added +44 to your rating, now it is equal to [+6862](http://www.netbout.com/b/35128?open=rating)
